flowmachine/flowmachine/core/hour_slice.py

```diff
@@ -0,0 +1,258 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import re
+from operator import lt as less_than, ge as greater_or_equal
+from typing import List, Union
+
+from sqlalchemy.sql import func, and_, or_, true
+
+__all__ = ["HourInterval", "HourSlice"]
+
+
+class DayPeriod:
+    """
+    Represents a repeated "daily" time period.
+    """
+
+    def __init__(self, weekday=None):
+        self.freq = "day"
+        self.weekday = weekday
+        if self.weekday is not None:
+            raise ValueError(
+                "If freq='day' then the `weekday` argument must not be provided."
+            )
+
+    def filter_timestamp_column_by_day_of_week(self, ts_col):
+        # no additional filtering needed to limit the day of the week
+        return true()
+
+
+class DayOfWeekPeriod:
+    """
+    Represents a repeated "day-of-the-week" time period.
+    """
+
+    valid_weekdays = [
+        "Monday",
+        "Tuesday",
+        "Wednesday",
+        "Thursday",
+        "Friday",
+        "Saturday",
+        "Sunday",
+    ]
+    weekday_indices_postgres = {
+        "Sunday": 0,
+        "Monday": 1,
+        "Tuesday": 2,
+        "Wednesday": 3,
+        "Thursday": 4,
+        "Friday": 5,
+        "Saturday": 6,
+    }
+
+    def __init__(self, weekday):
+        if weekday is None:
+            raise ValueError(
+                "If freq='week' then the `weekday` argument must be provided."
+            )
+        weekday = weekday.capitalize()
+        if weekday not in self.valid_weekdays:
+            valid_weekdays_list = ", ".join([repr(x) for x in self.valid_weekdays])
+            raise ValueError(
+                f"Invalid value for `weekday`. Must be one of: {valid_weekdays_list}."
+            )
+
+        self.freq = "week"
+        self.weekday = weekday
+
+    def filter_timestamp_column_by_day_of_week(self, ts_col):
+        return (
+            func.extract("dow", ts_col) == self.weekday_indices_postgres[self.weekday]
+        )
+
+
+def make_hour_interval_period(freq, *, weekday=None):
+    """
+    Returns an appropriate instance of `DayPeriod` or `DayOfWeekPeriod`,
+    depending on the value of `freq`.
+    """
+    cls_lookup = {"day": DayPeriod, "week": DayOfWeekPeriod}
+
+    try:
+        cls = cls_lookup[freq]
+    except KeyError:
+        allowed_freqs = ", ".join([repr(x) for x in cls_lookup.keys()])
+        raise ValueError(f"Argument `freq` must be one of: {allowed_freqs}.")
+
+    return cls(weekday=weekday)
+
+
+class HourOfDay:
```

Comment:
```suggestion
class HourOfDayPeriod:
```

For consistency with the others?

Reply:
Hmm, somehow this class doesn't seem on par to me with the other ones. `HourOfDay` represents a specific "timestamp" (or rather "hour-and-minutes-stamp"), i.e. a single point in time, whereas the other ones represent stretches / periods of time. But I'm the first one to admit that this is quite handwavy and the naming isn't very good yet (possibly also because there are perhaps even more meaningful concepts hidden here which should replace the existing ones).

Perhaps `HourOfDay` should be renamed to `HourStamp` or something similar? Any better suggestions?

Reply:
I have renamed this to `HourAndMinutesTimestamp`, but happy to use a better name if you can think of one.